bon-macros/src/util/mod.rs

```diff
@@ -45,6 +49,39 @@ where
     Ok((params, item))
 }
 
+pub(crate) fn parse_map_macro_input(
+    input: ParseStream<'_>,
+) -> Result<Punctuated<(Expr, Expr), Token![,]>, syn::Error> {
+    Punctuated::parse_terminated_with(input, parse_map_pair)
+}
+
+fn parse_map_pair(pair: ParseStream<'_>) -> Result<(Expr, Expr), syn::Error> {
+    let key = pair.parse().map_err(|_| pair.error(pair.to_string()))?;
+    let _: Token![:] = pair.parse()?;
+    let value = pair.parse()?;
+
+    Ok((key, value))
+}
+
+pub(crate) fn ensure_unique<'k, I>(keys: I) -> Result<()>
+where
+    I: IntoIterator<Item = &'k Expr>,
+{
+    let mut errors = Error::accumulator();
+
+    let mut exprs = std::collections::HashSet::new();
+
+    keys.into_iter().for_each(|key| {
+        if !exprs.insert(key.clone()) {
+            errors.push(err!(key, "duplicate map key"));
+        }
+    });
+
+    errors.finish()?;
+
+    Ok(())
+}
```

Comment:
Note that this check should not include expressions that include function or method calls, because functions are not guaranteed to be pure.

For example, in this check will reject the following valid code that would produce unique set items:

```rust
let i = 0;

let mut incr = || {
    i += 1;
    i
};

let set = bon::set![
    incr(),
    incr(),
    incr(),
];
```
In this case the set would've contained the values `1`, `2` and `3`, but the macro would reject it.

Instead we need to check only those expressions that consist of primitive literals `syn::ExprLit` and operators `+`, `-`, `==`, `!=`, etc. (not function calls).

Reply:
Didn't think of that one. We're now only checking for “pure” expressions (naming subject to bikeshedding) for uniqueness.